x/sync/keeper/change.go

```diff
@@ -0,0 +1,188 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/celer-network/sgn/x/sync/types"
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// SubmitChange create new change given a content
+func (keeper Keeper) SubmitChange(ctx sdk.Context, changeType string, data []byte, initiatorAddr sdk.AccAddress) (types.Change, error) {
+	changeID, err := keeper.GetChangeID(ctx)
+	if err != nil {
+		return types.Change{}, err
+	}
+
+	submitTime := ctx.BlockHeader().Time
+	votingPeriod := keeper.GetVotingParams(ctx).VotingPeriod
+	change := types.NewChange(changeID, changeType, data, submitTime, submitTime.Add(votingPeriod), initiatorAddr)
+
+	keeper.SetChange(ctx, change)
+	keeper.InsertActiveChangeQueue(ctx, changeID, change.VotingEndTime)
+	keeper.SetChangeID(ctx, changeID+1)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeSync,
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.ActionSubmitChange),
+			sdk.NewAttribute(types.AttributeKeyChangeID, fmt.Sprintf("%d", changeID)),
+		),
+	)
+
+	return change, nil
+}
+
+// ApproveChange adds a vote on a specific change
+func (keeper Keeper) ApproveChange(ctx sdk.Context, changeID uint64, voterAddr sdk.AccAddress) error {
+	change, ok := keeper.GetChange(ctx, changeID)
+	if !ok {
+		return sdkerrors.Wrapf(types.ErrUnknownChange, "%d", changeID)
+	}
+	if change.Status != types.StatusVotingPeriod {
+		return sdkerrors.Wrapf(types.ErrInactiveChange, "%d", changeID)
+	}
+
+	change.Voters = append(change.Voters, sdk.ValAddress(voterAddr))
+	keeper.SetChange(ctx, change)
+
+	return nil
+}
+
+// GetChange get change from store by ChangeID
+func (keeper Keeper) GetChange(ctx sdk.Context, changeID uint64) (change types.Change, ok bool) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := store.Get(types.ChangeKey(changeID))
+	if bz == nil {
+		return
+	}
+	keeper.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &change)
+	return change, true
+}
+
+// SetChange set a change to store
+func (keeper Keeper) SetChange(ctx sdk.Context, change types.Change) {
+	store := ctx.KVStore(keeper.storeKey)
+	bz := keeper.cdc.MustMarshalBinaryLengthPrefixed(change)
+	store.Set(types.ChangeKey(change.ChangeID), bz)
+}
+
+// IterateChanges iterates over the all the changes and performs a callback function
+func (keeper Keeper) IterateChanges(ctx sdk.Context, cb func(change types.Change) (stop bool)) {
+	store := ctx.KVStore(keeper.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.ChangesKeyPrefix)
+
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		var change types.Change
+		keeper.cdc.MustUnmarshalBinaryLengthPrefixed(iterator.Value(), &change)
+
+		if cb(change) {
+			break
```

Comment:
what's the logic of` if cb(change) { break } `?

Reply:
It is just a mechanism to stop the iteration